File: Python/BOJ/11054 가장 긴 바이토닉 부분 수열.py
import sys
N = int(sys.stdin.readline())
x = list(map(int, sys.stdin.readline().split()))
y = [1]*N

for i in range(1, N):
    for j in range(0, i):
        if x[j] < x[i]:
            y[i] = max(y[i], y[j]+1)

for m in range(1, N):
    for n in range(m):
        if x[n] > x[m]:
            y[m] = max(y[m], y[n]+1)

print(max(y))

File: Java/BOJ/11054 가장 긴 바이토닉 부분 수열.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine()," ");
        int[] x = new int[N];
        int[] y = new int[N];
        for(int i = 0 ; i < N; i++) {
        	x[i] = Integer.parseInt(st.nextToken());
        	y[i] = 1;
        }
        for (int i = 1; i < N; i++) {
        	for(int j = 0; j <i; j++) {
        		if(x[j] < x[i]) {
        			y[i] = Math.max(y[i], y[j]+1);
        		}
        	}
        }
        for (int i = 1; i < N; i++) {
        	for(int j = 0; j <i; j++) {
        		if(x[j] > x[i]) {
        			y[i] = Math.max(y[i], y[j]+1);
        		}
        	}
        }
        int maxValue = 0;
        for(int i = 0; i < N; i ++) {
        	if (maxValue < y[i]) {
        		maxValue = y[i];
        	}
        }
        System.out.println(maxValue);
	}
}